src/Microsoft.Health.Fhir.CosmosDb/Features/Metrics/CosmosMetricProcessor.cs

```diff
@@ -0,0 +1,119 @@
+        }
+
+        /// <summary>
+        /// Updates the request context with Cosmos DB info and updates response headers with the session token and request change values.
+        /// </summary>
+        /// <param name="sessionToken">THe session token</param>
+        /// <param name="responseRequestCharge">The request charge.</param>
+        /// <param name="collectionSizeUsageKilobytes">The size usage of the Cosmos collection in kilobytes.</param>
+        /// <param name="statusCode">The HTTP status code.</param>
+        public void ProcessResponse(string sessionToken, double responseRequestCharge, long? collectionSizeUsageKilobytes, HttpStatusCode? statusCode)
+        {
+            if (_fhirRequestContextAccessor.FhirRequestContext == null)
+            {
+                return;
+            }
+
+            if (!string.IsNullOrEmpty(sessionToken))
+            {
+                _fhirRequestContextAccessor.FhirRequestContext.ResponseHeaders[CosmosDbHeaders.SessionToken] = sessionToken;
+            }
+
+            AddRequestChargeToFhirRequestContext(responseRequestCharge, collectionSizeUsageKilobytes, statusCode);
+        }
+
+        private void AddRequestChargeToFhirRequestContext(double responseRequestCharge, long? collectionSizeUsage, HttpStatusCode? statusCode)
+        {
+            IFhirRequestContext requestContext = _fhirRequestContextAccessor.FhirRequestContext;
+
+            // If there has already been a request to the database for this request, then we want to append a second charge header.
+            if (requestContext.ResponseHeaders.ContainsKey(CosmosDbHeaders.RequestCharge))
+            {
+                var newHeaderValue = new StringValues(requestContext.ResponseHeaders[CosmosDbHeaders.RequestCharge].Append(responseRequestCharge.ToString(CultureInfo.InvariantCulture)).ToArray());
+                requestContext.ResponseHeaders[CosmosDbHeaders.RequestCharge] = newHeaderValue;
+            }
+            else
+            {
+                requestContext.ResponseHeaders[CosmosDbHeaders.RequestCharge] = responseRequestCharge.ToString(CultureInfo.InvariantCulture);
+            }
+
+            var cosmosMetrics = new CosmosStorageRequestMetricsNotification(requestContext.AuditEventType, requestContext.ResourceType)
+            {
+                TotalRequestCharge = responseRequestCharge,
+            };
+
+            if (collectionSizeUsage.HasValue)
+            {
+                cosmosMetrics.CollectionSizeUsageKilobytes = collectionSizeUsage;
+            }
+
+            if (statusCode.HasValue && statusCode == HttpStatusCode.TooManyRequests)
+            {
+                cosmosMetrics.ThrottledCount = 1;
+            }
+
+            cosmosMetrics.RequestCount = 1;
+
+            try
+            {
+                _mediator.Publish(cosmosMetrics, CancellationToken.None);
```

Comment:
For this try/catch to work you'll need to declare the method `async void` and `await` here. The method will return as soon as this is hit though, without waiting for publish to finish.

Reply:
Refactored to be `async`.